x/bank/internal/keeper/keeper.go

```diff
@@ -279,36 +274,47 @@ func (keeper BaseSendKeeper) AddCoins(ctx sdk.Context, addr sdk.AccAddress, amt
 		return nil, sdkerrors.Wrap(sdkerrors.ErrInvalidCoins, amt.String())
 	}
 
-	oldCoins := keeper.GetCoins(ctx, addr)
-	newCoins := oldCoins.Add(amt...)
+	var resultCoins sdk.Coins
 
-	if newCoins.IsAnyNegative() {
-		return amt, sdkerrors.Wrapf(
-			sdkerrors.ErrInsufficientFunds, "insufficient account funds; %s < %s", oldCoins, amt,
-		)
+	for _, coin := range amt {
+		oldBalance := keeper.GetBalance(ctx, addr, coin.Denom)
+		newBalance := oldBalance.Add(coin)
+		if newBalance.IsNegative() {
+			return nil, sdkerrors.Wrapf(
+				sdkerrors.ErrInsufficientFunds, "insufficient account funds; resulting balance %s is negative", newBalance,
+			)
+		}
+		keeper.SetBalance(ctx, addr, newBalance)
+		resultCoins = resultCoins.Add(newBalance)
 	}
 
-	err := keeper.SetCoins(ctx, addr, newCoins)
-	return newCoins, err
+	return resultCoins, nil
 }
 
-// SetCoins sets the coins at the addr.
+// SetCoins sets the balances for multiple denoms at the addr.
 func (keeper BaseSendKeeper) SetCoins(ctx sdk.Context, addr sdk.AccAddress, amt sdk.Coins) error {
+	for _, coin := range amt {
```

Comment:
Note: this is possibly *not* equivalent to the old logic, in that it won't set denoms not in `amt` to zero.

Reply:
Ah, you're right, this is different. I think the functionality of this function should be like this (with note in CHANGELOG), and a second function should be added to the keeper `ClearAllBalances(addr)`, which will set all balances to 0.

Then anyone who wants that original functionality would call clear all balances first, then setcoins

Reply:
Why not just call `ClearAllBalances(addr)` at the beginning of `SetCoins`?

Reply:
That's exactly what I'm saying, hence my `TODO`.